### ISSUE TITLE
'Zoom level 2' UI - considerations and tracking progress

### ISSUE THREAD

**Issue by Siequnu** [cyipt/actdev#37]
There are three broad 'zoom levels' in the ActDev project, national, site-level and agent level. This issue is about the 2nd zoom level. Considerations for the '2nd zoom level' UI:

- Integrate the two current study areas into a Mapbox instance, with toggles to enable each geojson layer currently present in the [data-small](https://github.com/cyipt/actdev/tree/main/data-small) folder
- Start to add data visualisation — different colours in buildings_od, line thicknesses, etc
- Enable certain additional CycleStreets layers that might be useful, i.e. collisions

With this there will essentially be two zoom levels — ❶ study area and ❷ individual study area with data. The UX & UI integration of the third ([abstreet](https://github.com/dabreegster/abstreet) layer) can then be considered. I'll probably start with a button, but there is a lot we could do to better integrate the various views.

Ported across from #11:

Data that needs to be generated - to be done after https://github.com/cyipt/actdev/milestone/3 is done:

- [ ] Routes to the site (fastest, balanced and quietest) 
- [ ] Cyclability ratings

Extras (not MVP items from a planning perspective but nice to have):

- [ ] Traffic count summaries, e.g. growth in traffic
- [ ] Road network summaries (e.g. % network that is 20 mph or less from OSM)

**Comment by Robinlovelace** [comment]
Isn't this a duplicate of #11 ?

**Comment by Robinlovelace** [comment]
I suggest closing #11 and moving the discussion here. Also there are Martin's comments in #19 :+1: 

> To reiterate, I think the technology approach cannot be decided on until we know what the functionality will be. There's lots above I could comment on, but again I don't think it's appropriate to speculate on technologies before user needs are properly discussed and prototyped.
> 
> > visualising datasets to change planning decisions so that they enable walking and cycling uptake and reduce car dependency associated with new developments
> 
> This is a very helpful summary of the outcomes/impact, but it is not a definition of functionality.
> 
> > UI specific aims has been discussed in #11.
> 
> Ticket 11 refers to a prototype of just showing data on a map. That does not imply a good tool ready for widespread use.
> 
> For instance, in CyIPT we ended up with a layer-based website which makes it easy to access that data, but I would still argue it is quite 'prototypey' because it doesn't really aim to guide the user towards a set of clear research questions - it just presents the data and assumes the users will somehow work out what's important. It contains a whole pile of buttons, each of which has equal visual prominence, when in fact some controls are actually for more important than others. A good UI should in my view be trying actively to orientate a user.
> 
> For instance, UI questions that I expect will arise during the UI prototyping phase might be:
> 
>     * What specific questions and angle should the user be able to ask? In other words, what actually is the functionality (as distinct from what data is available)?
> 
>     * Do we want a map which is nationally focussed (Bikedata, StreetFocus, CycleStreets, etc. or regionally focussed (PCT)?
> 
>     * Do we have a map that enables an initial orientation of the data but then allows people to move into a more specific interrogation system?
> 
>     * Do we start with a set of buttons that effectively form an entry point to wizards to guide people to a view?
> 
>     * Do we start a location search as a starting point (see FixMyStreet, StreetFocus, etc.) or just dump people directly into a UI?
> 
>     * Do we start with a map but have particular layers significantly more prominent?
> 
>     * Does the user actually click on a map for data to appear in a highly customised format, or is it pre-presented for browsing?
> 
>     * Does the map contain a set of controls to be operated in parallel or is something more linear wanted?
> 
> 
> These are all things we can't yet answer but will arise in the process of looking at what data we have, seeing what findings could be interrogable from it, and designing a user workflow.
> 
> If we just want to throw a load of data on a map, which is fine for merely interrogating the data in an alpha manner, I would probably go for Mapbox GL v1, and we already have experience in that, but I hope, having properly budgeted for UI development, we aren't ultimately planning to release a tool that is just a load of data thrown on a map :)

**Comment by Robinlovelace** [comment]
> Start to add data visualisation — different colours in buildings_od, line thicknesses, etc

I don't think the `buildings_od` needs to be in the UI, I just put them there to support reproducibility with respect to #34

**Comment by Robinlovelace** [comment]
Heads-up @Siequnu I think I realise what's going on now. #11 seems to be about the '1st zoom' level. This is about the 2nd zoom level, focussed on data that in (and to be put into) https://github.com/cyipt/actdev/tree/main/data-small , correct? I think it's worth closing #11 anyway as it was conflating two separate bits of work. Thoughts @mvl22 also very welcome on this. I think answers to Martin's questions in https://github.com/cyipt/actdev/issues/37#issuecomment-762889213 are important and will provide some preliminary answers in another comment.

**Comment by Robinlovelace** [comment]
Heads-up @mvl22 belated answer to your questions, that I think belong here:


For instance, UI questions that I expect will arise during the UI prototyping phase might be:

> What specific questions and angle should the user be able to ask? In other words, what actually is the functionality (as distinct from what data is available)?

I think the user should be able to ask at the site level:
   - Where is the development?
  - How big is it, in terms of area (simply showing its outline should help there) and n. houses?
  - Where are people likely to travel, based on existing data (the desire lines data should be able to answer this question)?
  - How are people likely to travel under the baseline scenario (the data in the desire-lines.geojson answers this I think)
  - How could people travel under a sustainable travel scenario (the `godutch` columns answer that question)?
  - What do JTS datasets tell us about the level of accessibility to the site?
  - Where are nearby trip attractors such as supermarkets (don't have data yet, see #15 )

> Do we want a map which is nationally focussed (Bikedata, StreetFocus, CycleStreets, etc. or regionally focussed (PCT)?

  - For this '2nd zoom level' UI I think StreetFocus level

> Do we have a map that enables an initial orientation of the data but then allows people to move into a more specific interrogation system?

  - Not sure, but don't think this affects the initial prototype. Thoughts @mvl22 ? A/B Street could be the 'more specific' system.

> Do we start with a set of buttons that effectively form an entry point to wizards to guide people to a view?

  - I suggest not in the first instance but open to suggestions, what do you think @mvl22 and @Siequnu ?
  
> Do we start a location search as a starting point (see FixMyStreet, StreetFocus, etc.) or just dump people directly into a UI?

  - Good question, I suggest jumping directly into the UI but no strong preference
  
> Do we start with a map but have particular layers significantly more prominent?

  - TBC - I think the site and desire lines and mode split data should be prominent as a starter for 10 but we can iterate on this - thoughts welcome
  
> Does the user actually click on a map for data to appear in a highly customised format, or is it pre-presented for browsing?

  - Pre-presented I think - would be good if they can download the geojson data but not a priority in the prototype
  
> Does the map contain a set of controls to be operated in parallel or is something more linear wanted?

  - The map should have controls, not 100% sure what is meant by this question but I think MapBox functionality should be fine

**Comment by mvl22** [comment]
Repo ready for writing into at:
https://github.com/cyipt/actdev-ui

**Comment by mvl22** [comment]
> #11 seems to be about the '1st zoom' level. This is about the 2nd zoom level, focussed on data that in (and to be put into) https://github.com/cyipt/actdev/tree/main/data-small , correct?

Yes, though sorry I should have suggested to Patrick we just continue with #11 as that could be both 1st and 2nd. But things have moved on anyway so this new ticket is clear enough.

**Comment by Robinlovelace** [comment]
All good, also provided a good motivation to create a issue on each zoom level. See now:

#17 
#28 
#37

Should be as easy as 1-2-3...

**Comment by Robinlovelace** [comment]
Nudge on this @mvl22 what do you think of my (updated) preliminary answers to your excellent spec questions above?

**Comment by mvl22** [comment]
> Nudge on this @mvl22 what do you think of my (updated) preliminary answers to your excellent spec questions above?

Yes, these are absolutely the kinds of responses we need to consider when building the final UI, and will be very useful for the workshop on Friday 29th.

Particularly the block starting "I think the user should be able to ask at the site level".

They are obviously less important at this stage for the prototype alpha stuff, which is really focussed on visualising the data and getting it on a map, so it can be scrutinised and ideas about how people might use the UI start to form. Clearly, turning those kinds of questions into a real end-user UI needs a lot more thinking than we will be doing at this alpha prototype UI phase of putting things onto the map, but they are excellent to have in mind.

**Comment by mvl22** [comment]
> Routes to the site (fastest, balanced and quietest)

In the first instance these will be separate files but later we will turn these into a single file with the route strategy type as an attribute, which will be better UI-wise as a filter property.

**Comment by mvl22** [comment]
NB There are currently two sites, as listed at
https://github.com/cyipt/actdev/tree/main/data-small

but there will be 35 areas in total eventually.

**Comment by Siequnu** [comment]
I’ll just integrate these two into the alpha UI as separate top level entities  — then on Friday we can begin to brainstorm accessible ways to display the information for a more numerous collection of sites.

**Comment by mvl22** [comment]
I've suggested to Patrick in a discussion earlier today we can include for now the place drop-down (like in CyIPT) to switch between sites in the alpha UI easily (though on CyIPT this was left unstyled).

@joeytalbot if it's easy, could you possibly create a single GeoJSON file at top-level in https://github.com/cyipt/actdev/tree/main/data-small which contains the boundary areas for each site. Obviously at present this would just show the two, but over time this will have all.

![Screenshot 2021-01-27 at 15 36 42](https://user-images.githubusercontent.com/361423/106014558-7f898c80-60b5-11eb-99d1-cba730061410.png)

**Comment by joeytalbot** [comment]
The site boundaries are already shown in https://github.com/cyipt/actdev/blob/main/all-sites.geojson

Does this file need to be moved into the data-small folder or is it fine as is?

**Comment by Robinlovelace** [comment]
> Does this file need to be moved into the data-small folder or is it fine as is?

From @mvl22's message above I think yes it's worth moving. Good to have all the data in one folder. It may be worth creating releases or a separate data repo for all 35 sites.

**Comment by Robinlovelace** [comment]
To avoid mixing code and data too much.

**Comment by joeytalbot** [comment]
Done

**Comment by mvl22** [comment]
Alpha UI screenshot at:
https://github.com/cyipt/actdev/issues/11

**Comment by mvl22** [comment]
We had a useful workshop on Friday morning.

Key outcomes and next steps:

[Alpha UI](https://actdev.cyipt.bike/) has really opened up the discussion about the data
Has been enabled us to do quick changes, e.g. colour, and start to understand the data as a whole
Final UI **not** expected to be the same as the Alpha map - needs a proper user-needs -based workflow
**Add further data and refine existing datasets in next 2-3 weeks**
**Then workshop to work on tailoring what user journey through the data**
Final UI work can then be done
Final UI pare down data heavily

**Comment by mvl22** [comment]
I'm pleased to say that the planning applications layer has now been migrated to a brand new CycleStreets API, just moved into beta 🎉.

This now makes that layer much faster (and therefore more usable) than the direct PlanIt API which we were hooking into before, which is not really designed for dynamic map browsing loads.

https://actdev.cyipt.bike/planningapplications/#14.3/52.17138/0.13034

Thanks to Andrew who supplied the enormous export of the dataset!

_Beta API limitations_: Note that it does not yet have source data updating enabled, so the data is essentially fixed to around the second week of January, but this will follow shortly. The date range selector in the UI is also not working but will fix shortly.

Now this layer is fast enough, I can now start to tweak it - I'm intending to set the application size to Medium,Large as the defaults. @joeytalbot - this should naturally mean that Outline/Full are likely to be shown naturally rather than them also needing to be filtered down to.

**Comment by mvl22** [comment]
Data awaited:

 - [ ] Generate 'dartboard' based on cycleability as concentric rings from development #46
 - [x] ~~Create spider-map from each site to enable visualisation of decay of cycling levels #47~~

**Comment by Robinlovelace** [comment]
>     * Generate 'dartboard' based on cycleability as concentric rings from development #46

A working prototype of this layer is available here (needs more attributes): https://github.com/cyipt/actdev/blob/main/data-small/great-kneighton/dartboard-1-3-6km.geojson

**Comment by Robinlovelace** [comment]
>     * Create spider-map from each site to enable visualisation of decay of cycling levels #47

I think that is a stretch target and not MVP.

**Comment by mvl22** [comment]
Action @Siequnu (as discussed earlier):

 - [x] Colour support for dartboard polygons
 - [x] Line thickness support bugfix

**Comment by mvl22** [comment]
> Key outcomes and next steps

Agreed today, the following timescales:

- w/c 8/Feb: Data work to ensure all datasets and fields present by UI workshop on 16/Feb, increase to at least 4 sites
- On 16/Feb: Workshop to work on tailoring what user journey through the data (9.30-11.30am).
- w/c 15/Feb: UI wireframing by Patrick; dependency: ALL datasets and fields present (even if some need refinement)
- w/c 22/Feb: UI week 1: Complete UI wireframing and start UI building (layout, design, coding) - Patrick (Martin to assist)
- w/c 29/Feb: UI week 2: UI building and bugfixing to fully-stable UI - Patrick (Martin to assist)
- w/c 8/Mar: UI week 3: Final bugfixes to polish for public workshop
- Data refinement to continue by Joey during the three UI weeks above
- On 12/Mar: Public workshop
- After workshop: Incorporate stakeholder feedback into data pipeline / UI, and continue to iterate

@Robinlovelace, @joeytalbot, @Siequnu.

**Comment by dabreegster** [comment]
One thing missing from that schedule: adding a button to open A/B Street for one of the sites? I hope from your end, it shouldn't be hard -- when the view is near one of the study areas, enable a button on the right panel. Upon clicking, open a URL like `http://abstreet.s3-website.us-east-2.amazonaws.com/dev/game/?--dev&cambridge/maps/trumpinton.bin&cam=18/52.20565/0.13871` (in a new tab?). The 3 details we could hash out:

- The `cam` parameter is the zoom/center lat/center lon that OSM uses
- The two map names right now are `cambridge/maps/trumpington` and `cheshire/maps/chapelford.bin`; maybe these can be inferred from existing study area names, or maybe we need to add a field, or maybe I need to rename things.
- We need to set up a deployment procedure to copy over the static files for abst to your webserver. I can give you a simple shell script to produce these (that would be slow the first time on a particular machine, due to installing Rust and the initial compile), or cut a release whenever we need and send you the files by Dropbox or something else.

Let me know if I should start a separate issue for this, or we can repurpose #28.

**Comment by Siequnu** [comment]
> Action @Siequnu (as discussed earlier):
> 
> * [x]  Colour support for dartboard polygons
> * [x]  Line thickness support bugfix

Both these have now been enabled.

In https://github.com/cyipt/actdev-ui/commit/5dc5d904256297dfe364965807d782c085612f17:
<img width="697" alt="Screenshot 2021-02-08 at 00 36 22" src="https://user-images.githubusercontent.com/11821606/107152979-156ec280-69a6-11eb-8ed9-55b8275ecef9.png">

<img width="834" alt="Screenshot 2021-02-08 at 00 36 12" src="https://user-images.githubusercontent.com/11821606/107152983-17388600-69a6-11eb-859f-925f5fb31ad8.png">

**Comment by mvl22** [comment]
> Let me know if I should start a separate issue for this, or we can repurpose #28.

Answered there, as that issue is precisely about integration with A/B Street so seems perfect.

**Comment by mvl22** [comment]
The multiple route types are now in place, as of https://github.com/cyipt/actdev-ui/commit/3cb758ddd359179c674a4be300a926d50dff98ab.

Sorry this took a while - it was a little more painful than expected because the form parameters are used to changing query strings, not strings with the base URL. Support for the latter is now in the upstream library. (NB there is is still some slightly buggy handling of URL persistency in drop-down forms, but that's a separate issue.)

![Screenshot 2021-02-15 at 22 46 43](https://user-images.githubusercontent.com/361423/107999638-cb0dc700-6fdf-11eb-9105-cf851e5cc02b.png)

**Comment by Robinlovelace** [comment]
My rough sketches from the meeting today:

![image](https://user-images.githubusercontent.com/1825120/108058839-81a79100-704c-11eb-904a-dc7c2a3b8b93.png)

**Comment by Robinlovelace** [comment]
Closing as over a week with no activity and to some extent the considerations have been bottomed out. Largely replaced by https://github.com/cyipt/actdev-ui/issues/13
